README
pyao - a Python wrapper module for the ao library

This is a wrapper for libao, an audio device abstraction
library. libao is available with ogg/vorbis at http://www.xiph.org.

To build you need distutils package from
http://www.python.org/sigs/distutils-sig/download.html (it comes with
Python 2.0). To install:

python config_unix.py
python setup.py build
[as root] python setup.py install

Use the config_unix.py script to configure the build first. You can
pass a --prefix argument to tell the script where you have the ao
files installed. If you have problems, check the file config.log for
specifics. If you have any problems let me know. Access the module by
using "import ao" in your Python code.

Here's an interactive session of just playing with the module, until I
create better documentation (there should be docstrings for
everything). Watch as I read some random data and "play" it to a wave
file.

>>> import ao

>>> dev = ao.AudioDevice('wav', filename = 'myoutput.wav')

>>> f = open('/dev/urandom', 'r') #that's some good stuff

>>> print dev
<AudioDevice object at 0x812ac28>

>>> print dev.driver_info()
{'author': 'Aaron Holtzman <aholtzma@ess.engr.uvic.ca>', 
 'short_name': 'wav', 
 'name': 'WAV file output', 
 'comment': 'Sends output to a .wav file'}

>>> print ao.driver_info('oss')
{'author': 'Aaron Holtzman <aholtzma@ess.engr.uvic.ca>', 
 'short_name': 'oss', 
 'name': 'OSS audio driver output ', 
 'comment': 'Outputs audio to the Open Sound System driver.'}

>>> data = f.read(1024*8)

>>> dev.play(data)

>>> <control-d>

And now I have a file myoutput.wav with random noise in it.

A note: Because of the way the AO API works, if you are opening a
device that outputs to a file (like raw or wav), then you HAVE to pass
the filename as a keyword parameter to the constructor (like
above). It can't just be an option (you used to be able to do
that). Opening a "live" device (e.g. oss or alsa), you obviously don't
have to worry about the filename.


Andrew Chatham <andrew.chatham@duke.edu>

ChangeLog
2015-05-24  Christian Schmitz <tynn.dev@gmail.com>
	* setup.py: Use setuptools if available

2015-02-12  Christian Schmitz <tynn.dev@gmail.com>
	* setup.py: Bumped version number to 1.2 to match libao
	* setup.py, aomodule.[ch]: Compatibility with Python 3
	* aomodule.c: Made use of converter for uint_64 in PyArg_ParseTuple*
	* aomodule.c: Set default_driver_id when driver_id is not given explicitly
	* aomodule.c: Added option matrix for ao_sample_format.matrix in parse_args
	* aomodule.[ch]: ao.AudioDevice is a type now
	* aomodule.h: Added driver_id member to ao.AudioDevice
	* aomodule.c: Assured num_bytes < #output_samples in ao.AudioDevice.play
	* aomodule.c: Let is_big_endian return bool
	* aomodule.c: Let driver_info return list of drivers for None argument
	* aomodule.[ch]: Added missing file_extension function and method
	* aomodule.c: Added constants AO_FMT_LITTLE, AO_FMT_BIG and AO_FMT_NATIVE
	* aomodule.c: Added macro AO_FE_MISSING for libao 1.1.0 compatibility

2004-01-13  Andrew H. Chatham <pyogg@andrewchatham.com>
	* aomodule.c: Add a patch from Roel Schroeven to add
	ao.default_driver_id()
	
2003-07-23  Andrew H. Chatham <pyogg@andrewchatham.com>
	* aomodule.c: (parse_args) and (py_ao_new) Apply patch from Joerg
	Lehmann to fix some segfaults.
	* aomodule.h: Apply patch from Klaus Harke to so things will
	compile on Suse properly.
	
2002-12-15  Andrew H. Chatham <pyogg@andrewchatham.com>
	* aomodule.c: ao_driver_id no longer accepts NULL as a default
	* aomodule.c: fixed some compiler warnings with gcc 3.2 
	
2002-9-24   Andrew H. Chatham <pyogg@andrewchatham.com>
	* setup.py: fixed regular expression
	
2001-12-09  Andrew H. Chatham <andrew.chatham@duke.edu>
	* README: Updated to document the "filename" keyword
	
	* config_unix.py: Updated to do logging.
	
2001-11-25  Andrew H. Chatham <andrew.chatham@duke.edu>
	* README: Updated to match the new API. I forgot.

	* src/aomodule.c: Changed get_driver_info as a method to driver_info

2001-09-02  Andrew H. Chatham <andrew.chatham@duke.edu>
	* setup.py: Bumped version number to 0.8 to match libao

	* config_unix.py (find_ao): pass the include dir to run_test
	
2001-08-30  Andrew H. Chatham <andrew.chatham@duke.edu>
	* aomodule.[ch] Updated to work with the updated ao API.
	  Not entirely sure the way of dealing with files is perfect,
	  but it should be serviceable.

	* aomodule.[ch] Changed some of the function names so that they
	  look more like their C counterparts (getting rid of get_ 
	  in function names, for example)
	
2001-02-18  Andrew H. Chatham <andrew.chatham@duke.edu>
	* README: Fixed incorrect config instructions
	
2001-01-31  Andrew H. Chatham <andrew.chatham@duke.edu>
	* setup.py, config_unix.py, setup.cfg: Went back to an older way
	of configuring things, which seems to be the only maintainable way
	to do it until distutils has better configuration
	support. "./setup.py bdist_rpm" should work fine now.
	
2001-01-21  Andrew H. Chatham <andrew.chatham@duke.edu>
	* src/aomodule.c (dict_to_options): Size checking removed; no
	longer necessary
	
	* src/aomodule.h: Made comments closer to Python style guide
        * src/aomodule.[ch]: Switched to C-style comments
	
	* src/aomodule.c (py_ao_new) (py_ao_get_driver_info): Support
	string argument

	* src/aomodule.c (initao): Removed extra calls to AddInt

	* ChangeLog: Started
	* build.py: removed
	
	* setup.py: Added configuration


test.py
#!/usr/bin/python

import ao

print "Default driver id:", ao.default_driver_id()
print "Bigendian?", ao.is_big_endian()
dev = ao.AudioDevice(ao.default_driver_id(), bits=16)
f = open('test.wav', 'r')
data = f.read()
dev.play(data, len(data))



config_unix.py
#!/usr/bin/env python

import string
import os
import sys

log_name = 'config.log'
if os.path.isfile(log_name):
    os.unlink(log_name)

def write_log(msg):
    log_file = open(log_name, 'a')
    log_file.write(msg)
    log_file.write('\n')
    log_file.close()

def exit(code=0):
    sys.exit(code)

def msg_checking(msg):
    print "Checking", msg, "...",

def execute(cmd):
    write_log("Execute: %s" % cmd)
    full_cmd = '%s 1>>%s 2>&1' % (cmd, log_name)
    return os.system(full_cmd)

def run_test(input, flags = ''):
    try:
        f = open('_temp.c', 'w')
        f.write(input)
        f.close()
        compile_cmd = '%s -o _temp _temp.c %s' % (os.environ.get('CC', 'cc'),
                                                  flags)
        write_log("executing test: %s" % compile_cmd)
        if not execute(compile_cmd):
            execute('./_temp')
    finally:
        execute('rm -f _temp.c _temp')
    
ao_test_program = '''
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ao/ao.h>

int main ()
{
  system("touch conf.aotest");
  return 0;
}
'''

def find_ao(ao_prefix = '/usr/local', enable_aotest = 1):
    """A rough translation of ao.m4"""

    ao_cflags = []
    ao_libs = []
    
    ao_include_dir = ao_prefix + '/include'
    ao_lib_dir = ao_prefix + '/lib'
    ao_libs = 'ao'

    msg_checking('for Ao')

    if enable_aotest:
        execute('rm -f conf.aotest')

        try:
            run_test(ao_test_program, "-I" + ao_include_dir)
            if not os.path.isfile('conf.aotest'):
                raise RuntimeError, "Did not produce output"
            execute('rm conf.aotest')
            
        except Exception, e:
            print "test program failed"
            return None

    print "success"

    return {'ao_libs' : ao_libs,
            'ao_lib_dir' : ao_lib_dir,
            'ao_include_dir' : ao_include_dir}

def write_data(data):
    f = open('Setup', 'w')
    for item in data.items():
        f.write('%s = %s\n' % item)
    f.close()
    print "Wrote Setup file"
            
def print_help():
    print '''%s
    --prefix                 Give the prefix in which ao was installed
                               (separated by a space)''' % sys.argv[0]
    exit()

def parse_args():
    def arg_check(data, argv, pos, arg_type, key):
        "Register an command line arg which takes an argument"
        if len(argv) == pos:
            print arg_type, "needs an argument"
            exit(1)
        data[key] = argv[pos]
        
    data = {}
    argv = sys.argv
    for pos in range(len(argv)):
        if argv[pos] == '--help':
            print_help()
        if argv[pos] == '--prefix':
            pos = pos + 1
            arg_check(data, argv, pos, "Ao Prefix", 'prefix')
    return data
    
def main():
    args = parse_args()
    prefix = args.get('prefix', '/usr/local')

    data = find_ao(ao_prefix = prefix)
    if not data:
        print "Config failure"
        exit(1)
    write_data(data)

if __name__ == '__main__':
    main()






src/aomodule.h
/*
 *  Copyright (C) 2001-2004  Andrew Chatham  <andrew.chatham@duke.edu>
 *  Copyright (C) 2015       Christian Schmitz  <tynn.dev@gmail.com>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#ifndef __AO_MODULE_H__
#define __AO_MODULE_H__

#include <Python.h>
#include <structmember.h>
#include <ao/ao.h>

#if PY_MAJOR_VERSION >= 3
#define PyInt_FromLong PyLong_FromLong
#define PyString_FromString PyUnicode_FromString
#endif

#define OPTSIZE 1024

typedef struct {
  PyObject_HEAD
  ao_device *dev;
  int driver_id;
} ao_Object;

static PyObject *Py_aoError;

static ao_option * dict_to_options(PyObject *);

static PyObject *py_ao_new(PyTypeObject *, PyObject *, PyObject *);

static void py_ao_dealloc(ao_Object *);

static char py_ao_play_doc[] =
"Play the contents of a given audio buffer.\n\
\n\
Arguments:\n\
buff : Buffer or string containing audio data\n\
n : Number of bytes to play (defaults to len(buff))";

static PyObject *py_ao_play(ao_Object *, PyObject *);

static char py_ao_driver_id_doc[] =
"Return the integer identifier for the driver with the given name (or object).";
static PyObject *py_ao_driver_id(PyObject *, PyObject *);

static char py_ao_driver_info_doc[] =
"Return a dictionary of information about a driver.\n\
\n\
It can either be called as a member function of an AudioDevice object:\n\
   x.driver_info()\n\
or as a standalone function which takes the integer id or the string name of the\n\
driver:\n\
   ao.driver_info(1) or ao.driver_info(\"pulse\")\n\
If None is passed or the argument omited a list of informations for all drivers\n\
is returned:\n\
   ao.driver_info() or ao.driver_info(None)";
static PyObject *py_ao_driver_info(PyObject *, PyObject *);

static char py_ao_is_big_endian_doc[] =
"Returns the endianness of the current host.";
static PyObject *py_ao_is_big_endian(PyObject *, PyObject *);

static char py_ao_doc[] =
"AudioDevice([driverid], bits=16, rate=44100, channels=2, byte_format=1, options=[], filename='', overwrite=0)\n\
OR\
AudioDevice(drivername, bits=16, rate=44100, channels=2, byte_format=1, options=[], filename='', overwrite=0)\n\
\n\
An AudioDevice object is an interface to a sound device. You can either pass\n\
an id (defaults to ao_default_driver_id) of a specific type of device or the\n\
name of that device type.\n\
If filename is passed, the module will try to open an output file as the\n\
audio device. In this case, overwrite indicates whether to overwrite an\n\
existing file\n";

static PyObject *py_ao_default_driver_id(PyObject *, PyObject *);
static char py_ao_default_driver_id_doc[] =
"Returns the ID number of the default live output driver.\n\
\n\
If the configuration files specify a default driver, its ID is returned,\n\
otherwise the library tries to pick a live output driver that will work\n\
on the host platform.\n\
\n\
Return values:\n\
    . a non-negative value is the ID number of the default driver\n\
    . -1 indicates failure to find a usable audio output device\n\
\n\
Notes:\n\
    If no default device is available, you may still use the \n\
    null device to test your application.";

static PyObject *py_ao_file_extension(PyObject *, PyObject *);
static char py_ao_file_extension_doc[] =
"Returns the normal file extension associated with a particular driver (like \"wav\" or \"au\").\n\
\n\
This is just an information function to allow library users to guess appropriate\n\
file names. You can safely ignore the recommended extension.\n\
\n\
Return Values:\n\
    . string containing suggested file extension.\n\
    . None if this driver has no file extension associated with it.";


struct PyMethodDef ao_Object_methods[] = {
  {"driver_info", py_ao_driver_info,
   METH_VARARGS, py_ao_driver_info_doc},
  {"file_extension", py_ao_file_extension,
   METH_VARARGS, py_ao_file_extension_doc},
  {"play", (PyCFunction) py_ao_play,
   METH_VARARGS, py_ao_play_doc},
  {NULL, NULL},
};

static PyMemberDef ao_Object_members[] = {
  {"driver_id", T_INT,
   offsetof(ao_Object, driver_id),
   READONLY, py_ao_driver_id_doc},
  {NULL},
};

static PyTypeObject ao_Type = {
  PyVarObject_HEAD_INIT(NULL, 0)
  "ao.AudioDevice",
  sizeof(ao_Object),
  0,
  (destructor) py_ao_dealloc,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  Py_TPFLAGS_DEFAULT,
  py_ao_doc,
  0,
  0,
  0,
  0,
  0,
  0,
  ao_Object_methods,
  ao_Object_members,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  py_ao_new,
};


struct PyMethodDef ao_methods[] = {
  {"driver_id", py_ao_driver_id,
   METH_VARARGS, py_ao_driver_id_doc},
  {"driver_info", py_ao_driver_info,
   METH_VARARGS, py_ao_driver_info_doc},
  {"is_big_endian", py_ao_is_big_endian,
   METH_VARARGS, py_ao_is_big_endian_doc},
  {"default_driver_id", py_ao_default_driver_id,
   METH_NOARGS, py_ao_default_driver_id_doc},
  {"file_extension", py_ao_file_extension,
   METH_VARARGS, py_ao_file_extension_doc},
  {NULL, NULL}
};

static char docstring[] =
"A Python wrapper for the ao library using in the ogg project.";


#endif /* __AO_MODULE_H__ */



src/aomodule.c
/*
 *  Copyright (C) 2001-2004  Andrew Chatham  <andrew.chatham@duke.edu>
 *  Copyright (C) 2015       Christian Schmitz  <tynn.dev@gmail.com>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "aomodule.h"
#include <assert.h>

static int
uint_32_obj (PyObject *obj, void *addr) {
  unsigned long value;
  uint_32 *value_32 = (uint_32 *) addr;

  value = PyLong_AsUnsignedLong(obj);

  if (value != -1 || !PyErr_Occurred()) {
    if (value <= 0xffffffff) {
      *value_32 = (uint_32) value;
      return 1;
    }

    PyErr_SetString(PyExc_OverflowError, "long int too large to convert");
  }

  return 0;
}

static ao_option *
dict_to_options(PyObject *dict)
{
  Py_ssize_t pos = 0;
  PyObject *key = NULL, *val = NULL, *tuple;
  ao_option *head = NULL;
  int ret = 1;
  const char *key_str, *val_str;

  if (!PyDict_Check(dict)) {
    PyErr_SetString(PyExc_TypeError, "Must be a dictionary");
    return NULL;
  }

  tuple = PyTuple_New(2);
  if (!tuple)
    return NULL;

  while (PyDict_Next(dict, &pos, &key, &val)) {
    PyTuple_SET_ITEM(tuple, 0, key);
    PyTuple_SET_ITEM(tuple, 1, val);

    if (!(ret = PyArg_ParseTuple(tuple, "ss", &key_str, &val_str)))
      break;

    if (!(ret = ao_append_option(&head, key_str, val_str))) {
      PyErr_SetString(Py_aoError, "Error appending options");
      break;
    }
  }

  Py_XINCREF(key);
  Py_XINCREF(val);
  Py_DECREF(tuple);

  if (ret)
    return head;

  ao_free_options(head);
  return NULL;
}

/*
   Helper function to parse everything out of the argument list.

   Returns 0 on failure.
*/
static int
parse_args(PyObject *args, PyObject *kwargs,
           ao_sample_format *format, PyObject **py_options,
           const char **filename,
           int *driver_id,
           int *overwrite)
{
  static const char *driver_id_kwlist[] = {"driver_id", "bits", "rate",
                                           "channels", "byte_format",
                                           "matrix", "options", "filename",
                                           "overwrite", NULL};
  static const char *driver_name_kwlist[] = {"driver_name", "bits",
                                             "rate", "channels",
                                             "byte_format", "matrix",
                                             "options", "filename",
                                             "overwrite", NULL};
  const char *driver_name = NULL;

  assert(py_options != NULL);
  assert(format != NULL);
  assert(filename != NULL);
  assert(driver_id != NULL);
  assert(overwrite != NULL);

  /* Set the default values */
  format->bits = 16;
  format->rate = 44100;
  format->channels = 2;
  format->byte_format = 1; /* What should this be by default? Anyone? */
  format->matrix = NULL;

  if(!(PyArg_ParseTupleAndKeywords(args, kwargs, "i|iiiisO!si",
                                   (char **) driver_id_kwlist,
                                   driver_id,
                                   &format->bits,
                                   &format->rate,
                                   &format->channels,
                                   &format->byte_format,
                                   &format->matrix,
                                   &PyDict_Type, py_options,
                                   filename, overwrite))) {
    PyErr_Clear();
    if(!(PyArg_ParseTupleAndKeywords(args, kwargs, "|ziiiisO!si",
                                   (char **) driver_name_kwlist,
                                   &driver_name,
                                   &format->bits,
                                   &format->rate,
                                   &format->channels,
                                   &format->byte_format,
                                   &format->matrix,
                                   &PyDict_Type, py_options,
                                   filename, overwrite)))
      return 0;

    if (driver_name)
      *driver_id = ao_driver_id(driver_name);
    else
      *driver_id = ao_default_driver_id();
  }

  return 1;
}

/*
  Actually create a new AudioDevice object
*/
static PyObject*
py_ao_new(PyTypeObject *type, PyObject *args, PyObject *kwargs)
{
  int driver_id, overwrite = 0;
  const char *filename = NULL;
  PyObject *py_options = NULL;
  ao_option *c_options = NULL;
  ao_device *dev;
  ao_sample_format sample_format;
  ao_Object *retobj;

  if (!parse_args(args, kwargs,
                  &sample_format, &py_options,
                  &filename, &driver_id, &overwrite))
    return NULL;

  if (py_options && PyDict_Size(py_options) > 0) {
    /* dict_to_options returns NULL on error, so you can't pass
       an empty dictionary. We can skip this then anyway. */

    c_options = dict_to_options(py_options);
    if (!c_options) {
      return NULL;
    }
  }

  if (filename == NULL) {
    dev = ao_open_live(driver_id, &sample_format, c_options);
  } else {
    dev = ao_open_file(driver_id, filename, overwrite,
                       &sample_format, c_options);
  }
  ao_free_options(c_options);

  if (dev == NULL) {
    PyErr_SetString(Py_aoError, "Error opening device.");
    return NULL;
  }

  retobj = PyObject_New(ao_Object, type);
  retobj->dev = dev;
  retobj->driver_id = driver_id;
  return (PyObject *) retobj;
}

static void
py_ao_dealloc(ao_Object *self)
{
  ao_close(self->dev);
  PyObject_Del(self);
}

static PyObject *
py_ao_driver_id(PyObject *self, PyObject *args)
{
  int driver_id;
  const char *str;

  if (!PyArg_ParseTuple(args, "s", &str))
    return NULL;

  driver_id = ao_driver_id(str);

  if (driver_id < 0) {
    PyErr_SetString(Py_aoError, "No such driver");
    return NULL;
  }

  return PyInt_FromLong(driver_id);
}

static PyObject *
info_to_obj(ao_info *info) {
  static char *keys[] = {"name", "short_name", "author", "comment"};
  PyObject *retdict;
  PyObject *string;
  char *strings[4];
  int i;

  if (!info) {
    PyErr_SetString(Py_aoError, "Error getting info");
    return NULL;
  }

  retdict = PyDict_New();
  if (retdict) {
    strings[0] = info->name;
    strings[1] = info->short_name;
    strings[2] = info->author;
    strings[3] = info->comment;

    for (i = 0; i < 4; i++) {
      string = PyString_FromString(strings[i]);
      if (!string || PyDict_SetItemString(retdict, keys[i], string) < 0) {
        Py_CLEAR(retdict);
        break;
      }
    }
  }

  return retdict;
}

static PyObject *
py_ao_driver_info(PyObject *self, PyObject *args)
{
  const char *driver_name = NULL;
  int driver_id = 0;
  ao_info **info_list;
  PyObject *retlist;
  PyObject *item;

  if (self != NULL && PyObject_TypeCheck(self, &ao_Type)) {

    /* It's a method */
    if (!(PyArg_ParseTuple(args, "")))
      return NULL;

    driver_id = ((ao_Object *) self)->driver_id;

  } else {

    /* Maybe it's a string, None or none */
    if ((PyArg_ParseTuple(args, "|z", &driver_name))) {
      if (driver_name) {

        driver_id = ao_driver_id(driver_name);

      } else {

        info_list = ao_driver_info_list(&driver_id);

        retlist = PyList_New(driver_id);
        if (retlist)
          while (driver_id--) {
            item = info_to_obj(info_list[driver_id]);
            if (!item || PyList_SetItem(retlist, driver_id, item) < 0) {
              Py_CLEAR(retlist);
              break;
            }
          }

        return retlist;
      }

    } else {

      /* Maybe it's an int */
      PyErr_Clear();
      if (!(PyArg_ParseTuple(args, "i", &driver_id)))
        return NULL;
    }
  }

  if (driver_id < 0) {
    PyErr_SetString(Py_aoError, "No such driver");
    return NULL;
  }

  return info_to_obj(ao_driver_info(driver_id));
}

static PyObject *
py_ao_file_extension(PyObject *self, PyObject *args)
{
  const char *driver_name;
  int driver_id;

  if (self != NULL && PyObject_TypeCheck(self, &ao_Type)) {

    /* It's a method */
    if (!(PyArg_ParseTuple(args, "")))
      return NULL;

    driver_id = ((ao_Object *) self)->driver_id;

  } else {

    /* Maybe it's a string, None or none */
    if ((PyArg_ParseTuple(args, "s", &driver_name))) {

      driver_id = ao_driver_id(driver_name);

    } else {

      /* Maybe it's an int */
      PyErr_Clear();
      if (!(PyArg_ParseTuple(args, "i", &driver_id)))
        return NULL;
    }
  }

  if (driver_id >= 0) {
#ifndef AO_FE_MISSING
    driver_name = ao_file_extension(driver_id);
    if (driver_name)
      return PyString_FromString(driver_name);
#endif

    if (ao_driver_info(driver_id))
      Py_RETURN_NONE;
  }

  PyErr_SetString(Py_aoError, "No such driver");
  return NULL;
}

static PyObject *
py_ao_play(ao_Object *self, PyObject *args)
{
  const char *samples;
  char *output_samples;
  uint_32 num_bytes = 0;
  int len;

#if PY_MAJOR_VERSION >= 3
  if (!(PyArg_ParseTuple(args, "y#|O&", &samples, &len, uint_32_obj, &num_bytes)))
#else
  if (!(PyArg_ParseTuple(args, "s#|O&", &samples, &len, uint_32_obj, &num_bytes)))
#endif
    return NULL;

  if (num_bytes == 0 || num_bytes > (uint_32) len)
    num_bytes = len;

  output_samples = PyMem_New(char, num_bytes);
  if (!samples)
    return PyErr_NoMemory();

  Py_BEGIN_ALLOW_THREADS
  Py_MEMCPY(output_samples, samples, sizeof(char) * num_bytes);
  ao_play(self->dev, output_samples, num_bytes);
  Py_END_ALLOW_THREADS

  PyMem_Del(output_samples);

  Py_RETURN_NONE;
}

static PyObject*
py_ao_is_big_endian(PyObject *self, PyObject *args)
{
  if (!PyArg_ParseTuple(args, ""))
    return NULL;
  if (ao_is_big_endian())
    Py_RETURN_TRUE;
  Py_RETURN_FALSE;
}

static PyObject*
py_ao_default_driver_id(PyObject *self, PyObject *args)
{
  /* Passed with METH_NOARGS */
  return PyInt_FromLong(ao_default_driver_id());
}


#define AddInt(x) PyModule_AddObject(module, #x, PyInt_FromLong(x));

#if PY_MAJOR_VERSION >= 3
#define RET_MOD (PyObject *)
static PyModuleDef aomodule = {
  PyModuleDef_HEAD_INIT, "ao", docstring,
  -1, ao_methods,
  NULL, NULL, NULL, NULL
};
PyMODINIT_FUNC
PyInit_ao(void)
#else
#define RET_MOD (void)
#ifndef PyMODINIT_FUNC  /* declarations for DLL import/export */
#define PyMODINIT_FUNC void
#endif
PyMODINIT_FUNC
initao(void)
#endif
{
  PyObject *module;

  if (PyType_Ready(&ao_Type) < 0)
    return RET_MOD NULL;

  Py_aoError = PyErr_NewException("ao.aoError", NULL, NULL);
  if (!Py_aoError)
    return RET_MOD NULL;

#if PY_MAJOR_VERSION >= 3
  module = PyModule_Create(&aomodule);
#else
  module = Py_InitModule3("ao", ao_methods, docstring);
#endif

  if (!module) {
    Py_DECREF(Py_aoError);
    return RET_MOD NULL;
  }

  Py_INCREF(&ao_Type);
  PyModule_AddObject(module, "AudioDevice", (PyObject *) &ao_Type);

  Py_INCREF(Py_aoError);
  PyModule_AddObject(module, "aoError", Py_aoError);

#ifdef AO_FMT_LITTLE
  AddInt(AO_FMT_LITTLE);
#endif

#ifdef AO_FMT_BIG
  AddInt(AO_FMT_BIG);
#endif

#ifdef AO_FMT_NATIVE
  AddInt(AO_FMT_NATIVE);
#endif

#ifdef AO_NULL
  AddInt(AO_NULL);
#endif

#ifdef AO_OSS
  AddInt(AO_OSS);
#endif

#ifdef AO_IRIX
  AddInt(AO_IRIX);
#endif

#ifdef AO_SOLARIS
  AddInt(AO_SOLARIS);
#endif

#ifdef AO_WIN32
  AddInt(AO_WIN32);
#endif

#ifdef AO_BEOS
  AddInt(AO_BEOS);
#endif

#ifdef AO_ESD
  AddInt(AO_ESD);
#endif

#ifdef AO_ALSA
  AddInt(AO_ALSA);
#endif

#ifdef AO_WAV
  AddInt(AO_WAV);
#endif

#ifdef AO_RAW
  AddInt(AO_RAW);
#endif

#ifdef AO_DRIVERS
  AddInt(AO_DRIVERS);
#endif

  ao_initialize();

  return RET_MOD module;
}

